File: horarium/calendar/prelent.js
function applyPrelent(metadata) {
	for (let w in metadata.prelent) {
		for (let d in metadata.prelent[w]) {
			let kyrie = d == 'sunday' ? 'common/kyrie/xvii.gabc' : 'common/kyrie/xviii.gabc';
			let collect = 'propers/prelent/' + w + '/collect.gabc'
			metadata.prelent[w][d].hours = {
				Vigils: {
					order: d == 'sunday' ? 'vigils/penitential-order-sunday.lit' : 'vigils/penitential-order-feria.lit',
					psalter: 'vigils/' + d + '.lit',
					propers: mergeDeep(vigils_commons(d), lessons('propers/prelent/' + w + '/' + d + '/vigils/'), {
						kyrie: kyrie,
						collect: collect
					}),
				},

				Lauds: {
					order: 'lauds/penitential-order.lit',
					psalter: 'lauds/' + d + '.lit',
					propers: mergeDeep(lauds_commons(d), {
						benedictus: 'propers/prelent/' + w + '/' + d + '/lauds/benedictus.lit',
						kyrie: kyrie,
						collect: collect
					}),
				},

				Prime: mergeDeep(minor_hours('prime', d), {
					propers: {
						collect: collect,
						kyrie: kyrie
					}
				}),

				Terce: mergeDeep(minor_hours('terce', d), {
					propers: {
						collect: collect,
						kyrie: kyrie
					}
				}),

				Sext: mergeDeep(minor_hours('sext', d), {
					propers: {
						collect: collect,
						kyrie: kyrie
					}
				}),
				
				None: mergeDeep(minor_hours('none', d), {
					propers: {
						collect: collect,
						kyrie: kyrie
					}
				}),

				Vespers: {
					order: 'vespers/penitential-order.lit',
					psalter: 'vespers/' + d + '.lit',
					propers: mergeDeep(vespers_commons(d), {
						magnificat: 'propers/prelent/' + w + '/' + d + '/vespers/magnificat.lit',
						kyrie: kyrie,
						collect: collect,
						chapter: (d == 'sunday' || d == 'saturday') ? 'propers/prelent/' + w + '/' + d + '/vespers/chapter.lit' : vespers_commons(d).chapter
					}),
				},

				Compline: {
					order: 'compline/penitential-order.lit',
					psalter: 'compline/ordinary.lit',
					propers: {
						hymn: 'hymn/te-lucis-ante-terminum.lit',
						chapter: 'common/compline/chapter(ordinary).lit',
						versicle: 'common/compline/chapter(ordinary).lit',
						canticle: 'common/compline/canticle(ordinary).lit',
						anthem: 'anthem/ave-regina-celorum.gabc',
						kyrie: kyrie,
						collect: 'common/compline/collect.gabc'
					}
				}
			}
		}

		metadata.prelent[w].sunday.hours.FirstVespers = metadata.prelent[w].saturday.hours.Vespers;
		metadata.prelent[w].sunday.hours.FirstCompline = metadata.prelent[w].saturday.hours.Compline;

		metadata.prelent[w].sunday.hours.FirstVespers.propers.kyrie = 'common/kyrie/xvii.gabc';
		metadata.prelent[w].sunday.hours.FirstCompline.propers.kyrie = 'common/kyrie/xvii.gabc';
	}
}

File: backend/src/liturgy/prelent.rs
use crate::kalendar::{Identifier, Season::*};
use crate::liturgy::Liturgy;
use std::collections::HashMap;
use std::path::PathBuf;
use chrono::{Weekday, Weekday::*};

pub fn resolve(iden: &Identifier) -> Liturgy {
    Liturgy {
        first_vespers: if "Sunday" == iden.day {
            Some(first_vespers(iden))
        } else {
            None
        },
        first_compline: if "Sunday" == iden.day {
            Some(compline(iden))
        } else {
            None
        },
        vigils: vigils(iden),
        matins: matins(iden),
        prime: prime(iden),
        terce: terce(iden),
        sext: sext(iden),
        none: none(iden),
        vespers: vespers(iden),
        compline: compline(iden),
    }
}

fn commons(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    let mut map: HashMap<&'static str, PathBuf> = HashMap::new();
    map.extend(crate::liturgy::commons::resolve(iden).unwrap());
    return map;
}

fn first_vespers(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    let mut map = HashMap::new();
    let day = iden.day.parse::<Weekday>().unwrap();

    map.insert("order", ["matins", "order", "penitential.lit"].iter().collect());

    map.insert("psalter", ["vespers", "saturday.lit"].iter().collect());

    map.insert("chapter", iden.to_path().join("vespers").join("chapter.lit"));

    map.insert(
        "hymn",
        [
            "hymn",
            "o-lux-beata-trinitas",
            match iden.season {
                PreLent(true) => "bvm.lit",
                PreLent(false) => "ordinary.lit",
                _ => panic!("Illegal season: {:?}", iden.season)
            },
        ]
        .iter()
        .collect(),
    );

    map.insert(
        "versicle",
        [
            "commons",
            "vespers",
            "versicles",
            if day == Sun {
                "sunday.lit"
            } else {
                "feria.lit"
            },
        ]
        .iter()
        .collect(),
    );

    map.insert(
        "canticle",
        iden.to_path().join("1st-vespers").join("magnificat.lit"),
    );

    map.insert("penitential-psalm", ["commons", "vespers", "penitential-psalm.lit"].iter().collect());
    
    map.extend(commons(iden));
    return map;
}

fn compline(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    let mut map = HashMap::new();

    map.insert("order", ["compline", "order", "penitential.lit"].iter().collect());

    map.insert("psalter", ["compline", "ordinary.lit"].iter().collect());

    map.insert(
        "chapter",
        ["commons", "compline", "chapters", "ordinary.lit"]
            .iter()
            .collect(),
    );

    map.insert(
        "hymn",
        ["hymn", "te-lucis-ante-terminum", match iden.season {
            PreLent(true) => "bvm.lit",
            PreLent(false) => "ordinary.lit",
            _ => panic!("Illegal season: {:?}", iden.season)
        }]
        .iter()
        .collect()
    );

    map.insert(
        "versicle",
        ["commons", "compline", "versicles", "ordinary.lit"]
            .iter()
            .collect(),
    );

    map.insert(
        "canticle",
        ["commons", "compline", "canticle", "ordinary.lit"]
            .iter()
            .collect(),
    );

    map.insert("anthem", [
        "commons",
        "compline",
        "anthems",
        match iden.season {
            PreLent(true) => "alma-redemptoris/christmas.lit",
            PreLent(false) => "ave-regina-caelorum.lit",
            _ => panic!("Illegal season: {:?}", iden.season)
        }
    ].iter().collect());

    map.extend(commons(iden));

    map.insert(
        "collect",
        ["commons", "compline", "collect.lit"].iter().collect(),
    );

    return map;
}

fn vigils(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    let mut map = HashMap::new();
    let vigils = crate::liturgy::commons::vigils(iden).unwrap();
    let day = iden.day.parse::<Weekday>().unwrap();

    map.extend(vigils);
    map.extend(commons(iden));
    map.extend(crate::liturgy::ordinary::resolve(iden).vigils);

    map.insert("order", ["vigils", "order", match day {
            Sun => "penitential-sunday.lit",
            _ => "penitential-feria.lit"
        }
    ].iter().collect());

    return map;
}

fn matins(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    HashMap::new()
}

fn prime(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    HashMap::new()
}

fn terce(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    HashMap::new()
}

fn sext(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    HashMap::new()
}

fn none(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    HashMap::new()
}

fn vespers(iden: &Identifier) -> HashMap<&'static str, PathBuf> {
    HashMap::new()
}
